refactor(linter/plugins): improve `Diagnostic` + `Suggestion` types

overlookmotel opened the pull request:
Improve `Diagnostic` and `Suggestion` types, by using `RequireAtLeastOne` utility type from `type-fest` package, to enforce that:

* Either `message` or `messageId` property must be provided.
* Either `node` or `loc` property must be provided.

Change to TSDown config `dts: { resolve: true }` is get it to bundle `type-fest` into the `.d.ts` file.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15643?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#15643** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15643?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15643?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#15642** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15642?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#15640** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15640?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#15639** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15639?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#15638** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15638?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=15643).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented:
@camc314 Can I ask for your help with this one?

It all seems good to me, except that we end up with this at the top of `dist/index.d.ts`:

```ts
import { RequireAtLeastOne } from "type-fest";
```

Presumably this won't work, as `type-fest` is not in the package's dependencies.

What's the conventional solution here?

* Add `type-fest` to dependencies? (seems not ideal, as it imposes the burden of downloading that package on users)
* Get TSDown to bundle the type into `index.d.ts`? (how?)

camc314 commented:
### Merge activity

* **Nov 12, 10:19 PM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Nov 12, 10:19 PM UTC**: `camc314` added this pull request to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc).
* **Nov 12, 10:24 PM UTC**: Merged by the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc).

overlookmotel committed:
refactor(linter/plugins): improve `Diagnostic` + `Suggestion` types (#15643)

Improve `Diagnostic` and `Suggestion` types, by using `RequireAtLeastOne` utility type from `type-fest` package, to enforce that:

* Either `message` or `messageId` property must be provided.
* Either `node` or `loc` property must be provided.

Change to TSDown config `dts: { resolve: true }` is get it to bundle `type-fest` into the `.d.ts` file.
